File: examples/python/lambda_update.py
"""Actualiza el código de una función Lambda existente y la vuelve a invocar.

Uso: python3 lambda_update.py <nombre-funcion>
"""
import io
import json
import sys
import zipfile

import boto3

lambda_client = boto3.client(
    "lambda",
    endpoint_url="http://localhost:4566",
    aws_access_key_id="test",
    aws_secret_access_key="test",
    region_name="us-east-1",
)

NUEVO_HANDLER = """
import json

def handler(event, context):
    return {
        "statusCode": 200,
        "body": json.dumps({"mensaje": "Versión actualizada del handler", "recibido": event}),
    }
"""


def build_zip():
    buffer = io.BytesIO()
    with zipfile.ZipFile(buffer, "w") as zf:
        zf.writestr("handler.py", NUEVO_HANDLER)
    return buffer.getvalue()


def main():
    if len(sys.argv) < 2:
        raise SystemExit("Uso: python3 lambda_update.py <nombre-funcion>")
    function_name = sys.argv[1]

    lambda_client.update_function_code(FunctionName=function_name, ZipFile=build_zip())
    print("Código actualizado para:", function_name)

    result = lambda_client.invoke(
        FunctionName=function_name,
        Payload=json.dumps({"prueba": "post-actualización"}).encode("utf-8"),
    )
    print("Respuesta tras actualizar:", result["Payload"].read().decode("utf-8"))


if __name__ == "__main__":
    main()

File: examples/java/LambdaUpdate.java
// Actualiza el código de una función Lambda existente y la vuelve a invocar.
// Uso: java LambdaUpdate <nombre-funcion>
import java.io.ByteArrayOutputStream;
import java.net.URI;
import java.nio.charset.StandardCharsets;
import java.util.zip.ZipEntry;
import java.util.zip.ZipOutputStream;
import software.amazon.awssdk.auth.credentials.AwsBasicCredentials;
import software.amazon.awssdk.auth.credentials.StaticCredentialsProvider;
import software.amazon.awssdk.core.SdkBytes;
import software.amazon.awssdk.regions.Region;
import software.amazon.awssdk.services.lambda.LambdaClient;

public class LambdaUpdate {
  private static final String NUEVO_HANDLER = """
      exports.handler = async (event) => {
        return {
          statusCode: 200,
          body: JSON.stringify({ mensaje: 'Versión actualizada del handler', recibido: event }),
        };
      };
      """;

  private static byte[] buildZip() throws Exception {
    ByteArrayOutputStream buffer = new ByteArrayOutputStream();
    try (ZipOutputStream zip = new ZipOutputStream(buffer)) {
      zip.putNextEntry(new ZipEntry("index.js"));
      zip.write(NUEVO_HANDLER.getBytes(StandardCharsets.UTF_8));
      zip.closeEntry();
    }
    return buffer.toByteArray();
  }

  public static void main(String[] args) throws Exception {
    if (args.length < 1) {
      throw new IllegalArgumentException("Uso: java LambdaUpdate <nombre-funcion>");
    }
    String functionName = args[0];

    LambdaClient lambda = LambdaClient.builder()
        .endpointOverride(URI.create("http://localhost:4566"))
        .region(Region.US_EAST_1)
        .credentialsProvider(
            StaticCredentialsProvider.create(AwsBasicCredentials.create("test", "test")))
        .build();

    try {
      lambda.updateFunctionCode(req -> req
          .functionName(functionName)
          .zipFile(SdkBytes.fromByteArray(buildZip())));
      System.out.println("Código actualizado para: " + functionName);

      var result = lambda.invoke(req -> req
          .functionName(functionName)
          .payload(SdkBytes.fromUtf8String("{\"prueba\":\"post-actualizacion\"}")));
      System.out.println("Respuesta tras actualizar: " + result.payload().asUtf8String());
    } finally {
      lambda.close();
    }
  }
}
